File: app/static/relatorio/secoes/ytd.js
/* Comparativo YTD — desenho de renderYTD() do Kit, alimentado pelo payload do servidor.
   Cada bloco só é desenhado se veio no payload (o servidor já tirou o que o perfil não vê). */
'use strict';
window.DESENHO=window.DESENHO||{};
DESENHO.ytd=function(P){
  let s='';
  const A=P['ytd.abertura'];
  if(A) s+=`<div class="sec-head"><div class="kick">Comparativo</div><h2>YTD Jan–${A.mLab} · ${A.Y0} vs ${A.Y1}</h2>
    <p class="lead">Período comparável janeiro a ${A.mLab.toLowerCase()}, venda contratada da base atualizada.</p></div>`;
  const K=P['ytd-kpi'];
  if(K){ const a=K.a,b=K.b,Y0=K.Y0,Y1=K.Y1;
    s+='<div class="kpis k3" data-blk="ytd-kpi">'+kpi(`Venda ${Y1} · Jan–`+K.mLab,mi(b.total),deltaHtml(b.total/a.total-1)+' vs '+Y0)
      +kpi(`Pedidos ${Y1}`,nf(b.peds),deltaHtml(b.peds/a.peds-1)+' vs '+Y0)
      +kpi(`Ticket médio ${Y1}`,mi(b.ticket),deltaHtml(b.ticket/a.ticket-1)+' vs '+Y0)+'</div>'; }
  const L=P['ytd-linha'];
  if(L) s+=fig(line([[String(L.Y0),L.serieA,SER[3]],[String(L.Y1),L.serieB,SER[0]]],
     L.labels,{valfmt:v=>mi(v,1),w:900,h:300}),ins('serieOscilacao'),'ytd-linha');
  const I=P['ytd-ind'];
  if(I){ const a=I.a,b=I.b;
    s+=B('ytd-ind',table(['Indicador',String(I.Y0),String(I.Y1),'Variação'],[
      ['Venda contratada',money(a.total),money(b.total),deltaHtml(b.total/a.total-1)],
      ['Pedidos',nf(a.peds),nf(b.peds),deltaHtml(b.peds/a.peds-1)],
      ['Quantidade',nf(a.qnt),nf(b.qnt),deltaHtml(b.qnt/a.qnt-1)],
      ['Ticket médio / pedido',money(a.ticket),money(b.ticket),deltaHtml(b.ticket/a.ticket-1)],
    ],['left','right','right','right'])); }
  const V=P['ytd-vendedor'];
  if(V){ const rows=V.linhas.map(x=>[esc(x[0]),money(x[1]),money(x[2]),
      `<span style="color:${x[2]-x[1]>=0?GOOD:BAD}">${money(x[2]-x[1])}</span>`, x[1]?deltaHtml(x[2]/x[1]-1):'<span class="mut">novo</span>']);
    s+=H3('Variação por vendedor(a)','','ytd-vendedor');
    s+=table(['Vendedor(a)',String(V.Y0),String(V.Y1),'Variação R$','Var. %'],rows,['left','right','right','right','right'],null,
      ins('rotatividade')); }
  return s;
};

File: app/calculo/secoes/ytd.py
# -*- coding: utf-8 -*-
"""Comparativo YTD — porte do cálculo de renderYTD() (app.js).

Devolve um payload por bloco, só com números e nomes; a formatação é do navegador."""
from ..datas import MES, ym_list
from ..vendas import PSEUDO_VEND, agregar


def calcular(C, params=None):
    DATA = C.blob('DATA')
    fim = C.maxym_vendas
    m_fim = fim % 100
    y1, y0 = C.ano_v, C.ano_v - 1
    i1, i0 = y1 * 100 + 1, y0 * 100 + 1
    a = agregar(DATA, i0, fim - 100)
    b = agregar(DATA, i1, fim)

    va = {}
    for x in a['vend']:
        va[x['name']] = x['v']
    vb = {}
    for x in b['vend']:
        vb[x['name']] = x['v']
    nomes = list(dict.fromkeys([x['name'] for x in a['vend']] + [x['name'] for x in b['vend']]))
    nomes = [n for n in nomes if n not in PSEUDO_VEND]
    linhas = [{'n': n, 'a': va.get(n, 0), 'b': vb.get(n, 0)} for n in nomes]
    linhas = [x for x in linhas if x['a'] + x['b'] > 0]
    linhas.sort(key=lambda x: -x['b'])

    resumo = lambda g: {'total': g['total'], 'peds': g['peds'], 'qnt': g['qnt'], 'ticket': g['ticket']}
    return {
        'ytd.abertura': {'Y0': y0, 'Y1': y1, 'mLab': MES[m_fim - 1]},
        'ytd-kpi': {'Y0': y0, 'Y1': y1, 'mLab': MES[m_fim - 1], 'a': resumo(a), 'b': resumo(b)},
        'ytd-linha': {'Y0': y0, 'Y1': y1, 'labels': MES[:m_fim],
                      'serieA': [a['monthly'].get(y, 0) for y in ym_list(i0, fim - 100)],
                      'serieB': [b['monthly'].get(y, 0) for y in ym_list(i1, fim)]},
        'ytd-ind': {'Y0': y0, 'Y1': y1, 'a': resumo(a), 'b': resumo(b)},
        'ytd-vendedor': {'Y0': y0, 'Y1': y1, 'linhas': [[x['n'], x['a'], x['b']] for x in linhas]},
    }
